lib/internal/promisify.js

```diff
@@ -0,0 +1,43 @@
+'use strict';
+
+module.exports = promisify;
+
+function promisify(fn, single) {
+  Object.defineProperty(inner, 'name', {
+    value: fn.name + 'Promisified'
+  });
+  return inner;
+
+  function inner() {
+    if (typeof arguments[arguments.length - 1] === 'function') {
+      return fn.apply(this, arguments);
+    }
+    var resolve = null;
+    var reject = null;
+    const args = Array.from(arguments);
+    const promise = new Promise((_resolve, _reject) => {
+      resolve = _resolve;
+      reject = _reject;
```

Comment:
Everything that follows can be nested in the Promise ctor, you will not need `_resolve`, `_reject` intermediate vars and you will be able to get rid of this awkward `resolve/reject` exposure out of the ctor scope.